### PR TITLE
libreoffice: wrapper rewrite

### PR DESCRIPTION
**Pull request by doronbehar**
###### Motivation for this change

So it will be possible for most users to use unwrapped `libreoffice`, see discussion at https://github.com/NixOS/nixpkgs/issues/141600

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 7c6f434c** [comment]
Kind of unfortunate we cannot move all the wrapping to wrapped (make cheap stuff not cause huge rebuilds), but Gapp/Qt wrappers need build inputs, right?

**Comment by 7c6f434c** [comment]
I am happy to have this amount of complexity in the wrapper (as iterations there have laughable overhead in comparison with the main expression), but if this is a list that needs maintaining, I am fine if you merge your version moving the presumably stable JAVA_HOME part into the main expression. Does some simple passthru trick let us do this thing mostly for free or not?

**Comment by doronbehar** [comment]
> but if this is a list that needs maintaining

Theoretically it does require maintenance, but I'm not sure practically it will. For instance now, opening the wrapped executable  in a text editor shows:

```
# ------------------------------------------------------------------------------------
# The C-code for this binary wrapper has been generated using the following command:


makeCWrapper '/nix/store/7ykydc4m2akw8pnliba2nqsqmi1994xp-libreoffice-7.3.7.2/bin/.soffice-wrapped' \
    --inherit-argv0 \
    --prefix 'GIO_EXTRA_MODULES' ':' '/nix/store/sarmffgvwncpdhgzrn631285f54wkpgy-dconf-0.40.0-lib/lib/gio/modules' \
    --prefix 'GIO_EXTRA_MODULES' ':' '/nix/store/sarmffgvwncpdhgzrn631285f54wkpgy-dconf-0.40.0-lib/lib/gio/modules' \
    --set 'GDK_PIXBUF_MODULE_FILE' '/nix/store/9vn32f9zhg0v5qkxvb09bkznrkdjv6lw-librsvg-2.55.1/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache' \
    --prefix 'XDG_DATA_DIRS' ':' '/nix/store/66mzgzydwls5h28kcnzrnjps054xdcgi-gsettings-desktop-schemas-43.0/share/gsettings-schemas/gsettings-desktop-schemas-43.0:/nix/store/0z70flbcxfvkp3qap9jdbi6mjjadwwkx-gtk+3-3.24.36/share/gsettings-schemas/gtk+3-3.24.36' \
    --prefix 'XDG_DATA_DIRS' ':' '/nix/store/7ykydc4m2akw8pnliba2nqsqmi1994xp-libreoffice-7.3.7.2/share' \
    --prefix 'GST_PLUGIN_SYSTEM_PATH_1_0' ':' '/nix/store/vlsjrglavg7xhrc7dn7ssxs7cnlg9jhr-gst-libav-1.20.3/lib/gstreamer-1.0:/nix/store/nfsb9n33mlb9hkyhnhsp14zdwfmzq3p2-gst-plugins-bad-1.20.3/lib/gstreamer-1.0:/nix/store/7sqcp998g235mw6dmw87i6lw28ch5fp1-gstreamer-1.20.3/lib/gstreamer-1.0:/nix/store/bam6j63fzcyq0sk9sw6xb3f9xizlq0pk-gst-plugins-base-1.20.3/lib/gstreamer-1.0:/nix/store/qbi6j713iaiachpnh815x4p539v7bhkw-gst-plugins-good-1.20.3/lib/gstreamer-1.0:/nix/store/dc0wgfa8402zfkgq3c0krj0xky90zcz2-gst-plugins-ugly-1.20.3/lib/gstreamer-1.0:/nix/store/vlsjrglavg7xhrc7dn7ssxs7cnlg9jhr-gst-libav-1.20.3/lib/gstreamer-1.0:/nix/store/nfsb9n33mlb9hkyhnhsp14zdwfmzq3p2-gst-plugins-bad-1.20.3/lib/gstreamer-1.0:/nix/store/7sqcp998g235mw6dmw87i6lw28ch5fp1-gstreamer-1.20.3/lib/gstreamer-1.0:/nix/store/bam6j63fzcyq0sk9sw6xb3f9xizlq0pk-gst-plugins-base-1.20.3/lib/gstreamer-1.0:/nix/store/qbi6j713iaiachpnh815x4p539v7bhkw-gst-plugins-good-1.20.3/lib/gstreamer-1.0:/nix/store/dc0wgfa8402zfkgq3c0krj0xky90zcz2-gst-plugins-ugly-1.20.3/lib/gstreamer-1.0' \
    --set-default 'JAVA_HOME' '/nix/store/jkv8vb24ab371hlsp55fj953kwh01m8n-openjdk-17.0.6+10'
```

And the list of dependencies there seem pretty stable to me. In the worst case, we need to rebuild the wrapper, which takes only a few seconds.

**Comment by 7c6f434c** [comment]
If all the options are considered cheap, I guess I would prefer single-wrapping-in-wrapper, so that if any new need for wrapping functionality arises we can add it easily. If not, I am fine with the current option as it is already tested.

**Comment by doronbehar** [comment]
The build takes ages :sweat: - the unwrapped derivation's `postInstall` required cleanup... I finally managed to build it and I think this is it.

**Comment by 7c6f434c** [comment]
Sorry to hear about postInstall… thanks for hunspell improvements in the wrapper!

@ofborg eval

**Comment by 7c6f434c** [comment]
I think the warning leaks into the evaluation and breaks the eval check (and probably would break the eval check on Hydra)

**Comment by doronbehar** [comment]
I found 3 usages of `libreoffice.libreoffice` in `all-packages.nix` that I missed before... Now I think this should be good, let's see.

File: pkgs/applications/office/libreoffice/default.nix [hunk 1]
```diff
@@ -102,15 +102,32 @@
...
+, ktextwidgets ? null
+, kwidgetsaddons ? null
+, kxmlgui ? null
+, phonon ? null
+, qtdeclarative ? null
+, qtquickcontrols ? null
+, qtsvg ? null
+, qttools ? null
+, solid ? null
+, sonnet ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/office/libreoffice/default.nix:130]: If we are refactoring things we should get rid of all those ``? null``. Those packages are not null and overwriting packages with null is not good style.

File: pkgs/applications/office/libreoffice/default.nix [hunk 2]
```diff
@@ -557,20 +564,56 @@ in
...
+      kwindowsystem
+      kxmlgui
+      phonon
+      qtbase
+      qtdeclarative
+      qtsvg
+      qttools
+      qtwayland
+      sonnet
+    ];
```
Comment (SuperSandro2000) [inline pkgs/applications/office/libreoffice/default.nix:616]: Why are passing through packages exactly which are not even used?

File: pkgs/applications/office/libreoffice/wrapper.nix [hunk 3]
```diff
@@ -1,33 +1,101 @@
...
+          HDIR="$PROFILE/share/hunspell"
+          if [ -d "$HDIR" ]; then
+              export DICPATH=$DICPATH''${DICPATH:+:}$HDIR
+          fi
+      done
+    '')
+  ] ++ lib.optionals dbusVerify [
+    # If no dbus is running, start a dedicated dbus daemon
+    "--run" (lib.escapeShellArg ''
+      if ! ( test -n "$DBUS_SESSION_BUS_ADDRESS" ); then
```
Comment (SuperSandro2000) [inline pkgs/applications/office/libreoffice/wrapper.nix:53]: I have the feeling that writing this into a script file will make things more manageable

File: pkgs/applications/office/libreoffice/wrapper.nix [hunk 4]
```diff
@@ -1,33 +1,101 @@
...
-  inherit coreutils dbus gnugrep libreoffice bash;
+
+let
+  makeWrapperArgs = builtins.concatStringsSep " " ([
+    "--set" "GDK_PIXBUF_MODULE_FILE" "${librsvg}/${gdk-pixbuf.moduleDir}.cache"
+    "--prefix" "GIO_EXTRA_MODULES" ":" "${lib.getLib dconf}/lib/gio/modules"
+    "--prefix" "XDG_DATA_DIRS" ":" "${unwrapped.gtk3}/share/gsettings-schemas/${unwrapped.gtk3.name}"
+    "--prefix" "XDG_DATA_DIRS" ":" "$out/share"
+    "--prefix" "XDG_DATA_DIRS" ":" "${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}"
+    "--prefix" "XDG_DATA_DIRS" ":" "${hicolor-icon-theme}/share"
```
Comment (SuperSandro2000) [inline pkgs/applications/office/libreoffice/wrapper.nix:27]: Does makeWrapper generate less shell if we combine this?

File: pkgs/applications/office/libreoffice/default.nix (hunk 2 above)
Comment (7c6f434c) [reply pkgs/applications/office/libreoffice/default.nix:616]: What do you mean not used? These are the packages used for generating the paths in the wrappers.

File: pkgs/applications/office/libreoffice/default.nix (hunk 1 above)
Comment (doronbehar) [reply pkgs/applications/office/libreoffice/default.nix:130]: And without the `? null` the expression will not evaluate.

File: pkgs/applications/office/libreoffice/default.nix (hunk 2 above)
Comment (doronbehar) [reply pkgs/applications/office/libreoffice/default.nix:616]: > Why are passing through packages exactly which are not even used?

It's a good question. I took all of the references I found in the original wrapped executable of `libreoffice-qt` that came from the `QT_PLUGIN_PATH` and the `QML2_IMPORT_PATH` environment variables. I'm not sure exactly why all of these dependencies came due to `wrapQt`, but I didn't want to brake anything for any `libreoffice-qt` user out there due to this change.

> What do you mean not used? These are the packages used for generating the paths in the wrappers.

I think @SuperSandro2000 meant that they are not used as dependencies.

File: pkgs/applications/office/libreoffice/wrapper.nix (hunk 4 above)
Comment (doronbehar) [reply pkgs/applications/office/libreoffice/wrapper.nix:27]: > Does makeWrapper generate less shell if we combine this?

Yes. But on the other hand, reading the generated wrapper is easier IMO when each environment variable has it's own 5 lines of shell variables assignments. If the maintainer prefers these environment variables combined, I won't mind change it.

File: pkgs/applications/office/libreoffice/wrapper.nix (hunk 3 above)
Comment (doronbehar) [reply pkgs/applications/office/libreoffice/wrapper.nix:53]: > I have the feeling that writing this into a script file will make things more manageable

I think this is more elegant this way. In any case, I think this should also be the maintainer's decision.

**Comment by doronbehar** [comment]
Thanks!

**Comment by raboof** [comment]
Since this change there no longer is a `libreoffice` symlink to `soffice`, and the LibreOffice applications no longer appear to show up in the Gnome applications drawer. I opened #220693 to track those

**Pull request merged**